### ISSUE TITLE
Creating large usdz files produces invalid files without error

### ISSUE THREAD

**Issue by ShaddyDC** [PixarAnimationStudios/USD#3098]
### Description of Issue

Trying to create a usdz file from a large usd folder with `UsdUtilsCreateNewUsdzPackage` creates an invalid file.
Opening it gives [this error](https://github.com/PixarAnimationStudios/OpenUSD/blob/2864f3d04f396432f22ec5d6928fc37d34bb4c90/pxr/usd/usd/crateFile.cpp#L3299), both in omniverse and usdview.
![image](https://github.com/PixarAnimationStudios/OpenUSD/assets/18403034/2801993c-835a-48c1-be53-502821bf44e9)

The file is not a valid zip file anymore either.
We suspect that the issue is that the file size is over the 32 bit limit and the ZIP64 extension is not supported, which is fine, but it would be nice if we then received an error code.
Actually being able to create those large archives would be even better, of course.

I can provide some concrete reproduction files, but given the file size we're talking about and the suspected cause, I figured it's probably not worth it.
I can also give a concrete code snippet to reproduce if needed.

### Steps to Reproduce
1. Create a large, valid folder that can be turned into an usdz archive
2. Call `UsdUtilsCreateNewUsdzPackage(path, outPath)` on the folder, which will return `true`
3. Ensure the created usdz is over 4 GB. If not, go back to step 1
4. Try to open the created usdz file with `usdview` or `unzip`, both of which fail

### System Information (OS, Hardware)
Replicated both on debian-bookworm in docker and on NixOS `24.05.20240515.33d1e75`.

### Package Versions

23.05, 23.11

### Build Flags

```
    "-DPXR_BUILD_ALEMBIC_PLUGIN=ON"
    "-DPXR_BUILD_DRACO_PLUGIN=ON"
    "-DPXR_BUILD_EMBREE_PLUGIN=ON"
    "-DPXR_BUILD_EXAMPLES=OFF"
    "-DPXR_BUILD_IMAGING=ON"
    "-DPXR_BUILD_MONOLITHIC=ON"
    "-DPXR_BUILD_TESTS=OFF"
    "-DPXR_BUILD_TUTORIALS=OFF"
    "-DPXR_BUILD_USD_IMAGING=ON"
    "-DPXR_BUILD_DOCUMENTATION=OFF"
    "-DPXR_BUILD_PYTHON_DOCUMENTATION=OFF"
    "-DPXR_BUILD_USDVIEW=OFF"
    "-DPXR_BUILD_USD_TOOLS=OFF"
    "-DPXR_ENABLE_MATERIALX_SUPPORT=OFF"
    "-DPXR_ENABLE_OSL_SUPPORT=OFF"
```

**Comment by jesschimein** [comment]
Filed as internal issue #USD-9711

**Comment by dgovil** [comment]
Yeah this is a good callout. I don't think there are any plans to extend USDZ to support zip64, but perhaps someone can add a warning like you suggested if any individual file or the total file size is going to be more than the 32 bit range.

**Comment by ShaddyDC** [comment]
Thanks for looking into the issue! It would be very useful for us if there were support in USDZ for large files like this in the future, as we regularly work with 3D data sets on that scale and want to offer them to users in the USD ecosystem.
Having a warning or preferably have the function fail would help a lot in working around it already, however.

**Comment by asluk** [comment]
> Having a warning or preferably have the function fail would help a lot in working around it already, however.

From my naive perspective-- could you add a file size check in your pipeline after the call to write out the USDZ?

**Comment by ShaddyDC** [comment]
Yes, that's probably the best option at the moment, but it doesn't feel like that should be the place to address this problem, more like a band-aid. Thank you for the suggestion though!

**Comment by asluk** [comment]
for sure-- not suggesting it as any more of a bandaid, but wanted to get a sense of how blocking of an issue this is-- thanks!

**Comment by nvmkuruc** [comment]
This appears to still be an issue when using `UsdUtils`.  This should create a `usda` file that's a little bigger than 4GB.

```
import random
import string
import sys

from pxr import Tf
from pxr import Sdf
from pxr import UsdUtils

layer = Sdf.Layer.CreateNew("./test_zip.usda")
for i in range(100):
    print(f"prim {i}")
    for j in range(100):
        attribute_spec = Sdf.CreatePrimAttributeInLayer(layer, f"/prim_{i}.attr_{j}", Sdf.ValueTypeNames.String)
        attribute_spec.default = ''.join(random.choice(string.ascii_letters) for _ in range(500000))
layer.Save()

# Close the layer. Make sure the source layer was valid and saved correctly.
del layer
assert(not Sdf.Layer.Find('./test_zip.usda'))

# Let `UsdUtils` convert the source file to `usdz`.
assert(UsdUtils.CreateNewUsdzPackage(Sdf.AssetPath("./test_zip.usda"), "./out.usdz"))

# Try to open the `usdz` layer.
try:
    package_layer = Sdf.Layer.FindOrOpen('./out.usdz')
except Tf.ErrorException as e:
    print(f"Invalid package layer: {e}", file=sys.stderr)
```

**Comment by spiffmon** [comment]
Hmmm.... @dgovil , any thoughts on what might be lacking in PR #3618 ?

**Comment by dgovil** [comment]
Not off the top of my head. I can look into it after the holiday break. I'm surprised since there's a check at the very end of the write out for size as well.

**Comment by dgovil** [comment]
Though just a note that what was integrated is only based on my PR and the actual implementation by @matthewcpp  is more extensive and different. I haven't validated that implementation myself.

**Comment by spiffmon** [comment]
Ah sorry, I got confused.  We'll have a look then!

--spiff

**Comment by drwave** [comment]
I think I’ve weighed in on this before, but I’ll just chime in with my 2 cents:

The fact that USDZ is 32 bit limited is…

Bad. Short sighted. Silly. Unfortunate. <insert many other negative, diminutive adjectives>

USDZ is a wonderful delivery component  of the USD ecosystem that allows one to package up complex multi-layer, super large w/EXR UDIM textures without caring that they have (literally XXX) thousands of files in there.

All that said, I know that I’ve raised this issue a bunch of times over the years, and I expect that smarter folks than I have raised reasonable issues that make my feeble attempts obviously wrong.

But! Since it keeps coming up…

I would love to see those raised issues once again to remind me of why my concerns (umm, everything is getting bigger?) are not relevant.

Thank you, I’ll take my answer off the air...

- wave

**Comment by matthewcpp** [comment]
Thanks for reporting this @nvmkuruc 

As @dgovil mentioned, the route that was chosen to address the reported issue was different then the PR they put forth in #3618.

The case that handled adding a file greater than 4gb which that PR addressed was not adequately covered in what was committed.  I have have a fix for that issue on our side.  Thanks for the handy repo script.

**Comment by spiffmon** [comment]
@drwave , you can have your answer on air!  There is no technical limitation or gigantic concern on the OpenUSD side for having 64-bit zip support.  In the desire to gt the original feature out, and given that 64 bit was an extension to the standard (yes, also decades old, now), we just decided to focus on getting the 32 bit support out first, and nobody blinked for six years :-) .  At this point it's a matter of resources/priorities.

> I would love to see those raised issues once again to remind me of why my concerns (umm, everything is getting bigger?) are not relevant.

I do want to call out we've been working with your colleagues to keep squeezing things smaller, e.g. AVIF and AOM geometry compression, but admittedly that doesn't help with production-grade environments with high-def geometry and exr textures.

**Comment by spiffmon** [comment]
And finally, can I suggest that with @matthewcpp 's upcoming change, we close this ticket as resolved since it will address the reported Issue, and file a new one that is a specific request for making 64 bit support official, so it can start being tracked and prioritized?

**Comment by dgovil** [comment]
@matthewcpp Just checking if there were any updated commits I should track in the dev branch for this to test with?

**Comment by matthewcpp** [comment]
@dgovil The fix for the issue reported by @nvmkuruc is e00c2fa7d0fded6aa376d785b6142dd9812ceaeb

Let me know if there are any other issues you encounter on your end.

**Comment by dgovil** [comment]
Ah thank you!

**Comment by dgovil** [comment]
Looks to be working here. @nvmkuruc idk if you wanna give it a whirl and then close this out.

**Comment by spiffmon** [comment]
Following my earlier suggestion, closing this in favor of Issue #4201 for further work.
